### PR TITLE
Cloudwatch Trail logging support

### PR DESCRIPTION
**Pull request by mmihic**

### CODE REVIEW

File: misk/src/main/kotlin/misk/cloud/aws/logging/CloudwatchLogService.kt [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+internal class CloudwatchLogService(
+  private val events: BlockingQueue<CloudwatchLogEvent>,
+  moshi: Moshi,
+  config: CloudwatchLogConfig,
+  logs: AWSLogs,
+  appName: String,
+  clock: Clock,
+  maxBatchByteSize: Int = DEFAULT_MAX_BATCH_BYTES_SIZE
+) : AbstractIdleService(), Runnable {
```
[the remark this answers is not in the document]
Comment (mmihic) [reply misk/src/main/kotlin/misk/cloud/aws/logging/CloudwatchLogService.kt]: It's ``run`` by the thread started in ``start()``

File: misk/src/main/kotlin/misk/cloud/aws/logging/CloudwatchLogService.kt [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        // what environment it would be necessary to do so)
+        retry(5, ExponentialBackoff(Duration.ofMillis(20), Duration.ofMillis(1000))) {
+          try {
+            val orderedEvents = events.sortedBy { it.timestamp }
+            val request = PutLogEventsRequest(logGroupName, logStreamName, orderedEvents)
+                .withSequenceToken(nextSequenceToken)
+            val result = logs.putLogEvents(request)
+            nextSequenceToken = result.nextSequenceToken
+          } catch (e: InvalidSequenceTokenException) {
+            nextSequenceToken = e.expectedSequenceToken
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/cloud/aws/logging/CloudwatchLogService.kt]: should we log something in the next batch?

File: misk/src/main/kotlin/misk/cloud/aws/logging/CloudwatchTrail.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  @VisibleForTesting
+  internal fun registerAppender(
+    config: CloudwatchLogConfig,
+    eventQueue: BlockingQueue<CloudwatchLogEvent>
+  ) {
+    val rootLogger = LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME) as Logger
+    val context = rootLogger.loggerContext
+    context.reset()
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/cloud/aws/logging/CloudwatchTrail.kt]: what is going on here?

File: misk/src/main/kotlin/misk/cloud/aws/logging/CloudwatchTrail.kt [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import misk.environment.InstanceMetadata
+import org.slf4j.LoggerFactory
+import java.time.Clock
+import java.util.concurrent.ArrayBlockingQueue
+import java.util.concurrent.BlockingQueue
+
+/** Enables logging for Cloudwatch Trail */
+object CloudwatchTrail {
+  /** Enable Cloudwatch Trail logging for the given app */
+  fun enableLogging(
```
Comment (ryanhall07) [inline misk/src/main/kotlin/misk/cloud/aws/logging/CloudwatchTrail.kt]: where does this get enabled?

Comment (mmihic) [reply misk/src/main/kotlin/misk/cloud/aws/logging/CloudwatchTrail.kt]: Right now it has to be explicitly called from the top level application code of services running in AWS. Ultimately though I want to have something like an ``installCloudSupport`` method that figures out what cloud you are on and installs all of the support infra to tie into that cloud's environmental systems.

**Pull request merged**